include/flatbuffers/flatbuffers.h

```diff
@@ -557,15 +562,56 @@ class vector_downward {
 
  private:
   // You shouldn't really be copying instances of this class.
-  vector_downward(const vector_downward &);
-  vector_downward &operator=(const vector_downward &);
+  vector_dir(const vector_dir &);
+  vector_dir &operator=(const vector_dir &);
 
   size_t reserved_;
   uint8_t *buf_;
   uint8_t *cur_;  // Points at location between empty (below) and used (above).
   const simple_allocator &allocator_;
 };
 
+template<> inline void
+vector_dir<UPWARD>::clear() {
+  if (buf_ == nullptr)
+    buf_ = allocator_.allocate(reserved_);
+
+  cur_ = buf_;
+}
+
+template <> inline uoffset_t
+vector_dir<UPWARD>::size() const {
+  assert(cur_ != nullptr && buf_ != nullptr);
+  return static_cast<uoffset_t>(cur_ - buf_);
+}
+
+template <> inline uint8_t *
+vector_dir<UPWARD>::make_space(size_t len) {
+  if (len > static_cast<size_t>(cur_ - buf_)) {
+    auto old_size = size();
+    auto largest_align = AlignOf<largest_scalar_t>();
+    reserved_ += (std::max)(len, growth_policy(reserved_));
+    // Round up to avoid undefined behavior from unaligned loads and stores.
+    reserved_ = (reserved_ + (largest_align - 1)) & ~(largest_align - 1);
+    auto new_buf = allocator_.allocate(reserved_);
+    auto new_cur = new_buf + old_size;
+    memcpy(new_buf, buf_, old_size);
+    cur_ = new_cur;
+    allocator_.deallocate(buf_);
+    buf_ = new_buf;
+  }
+  cur_ += len;
+  // Beyond this, signed offsets may not have enough range:
+  // (FlatBuffers > 2GB not supported).
+  assert(size() < FLATBUFFERS_MAX_BUFFER_SIZE);
+  return cur_;
```

Comment:
How does this work? This function returns a pointer to write to. It points to above the new space, not the start of the new space.